docs/proposals/moving-to-dynamodb.md

```diff
@@ -0,0 +1,358 @@
+servers running [Cassandra] instances. This design change is an effort to both
+ease the setup of new `warehouse.ai` instances and provide some cost-savings
+for instance-maintainers. This will be the first step to transform
+`warehouse.ai` to more closely align with modern `serverless` style
+architecture. In the short-term, we will be specifically targeting [DynamoDB],
+but welcome [contributions][contributing] to be able to expand the scope to
+meet other data stores.
+
+## Existing instances & data migration
+
+If you have an existing running instance of `warehouse.ai` and have concerns,
+we would love to hear from you and help you with any data migration.
+
+## Data Models
+
+The current data models and [Cassandra] table design for `warehouse.ai`
+lives in [`warehouse-models`] and [`warehouse.ai-status-models`]. They are
+summarized here:
+
+* `build` - Represent an individual build of a package
+* `build_file` - Represent an individual file (unit)
+* `build_head` - Represent the head build version of an entire package
+* `dependent` - A dependency graph where every packaged publish can ensure
+that any package that depends on it can be updated
+* `dependent_of` - An inverse of `dependent` in order for a dependent package
+to see what its parent is
+* `release_line` - Represent all the necessary information for a given
+package/version to know what needs to be deployed
+* `release_line_dep` - Represent all the necessary information for a given
+package/version to know what needs to be deployed
+* `release_line_head` - Represents the head release-line for a given package.
+* `version` - Records of every npm publish of a package to the registry
+* `package` - Represent an entire published packaged `package.json` to the
+registry
+* `package_cache` - Stores the same data as `package`, but forced into a single
+partition
+* `status` - Generic status information
+* `status_head` - Generic status information but just for the latest version
+for a given `pkg` and `env`.
+* `status_event` - A detailed event for the various stages of a build process
+* `status_counter` - A simple distributed counter model that is incremented
+when a `locale` build has completed in order to compute progress
+
+These tables should for the most part just migrate into DynamoDB tables without
+significant changes. The largest changes will be needing to create our own
+composite keys since DynamoDB doesn't support them.
+
+DynamoDB also doesn't support the Cassandra `keyspace` mechanism, so each table
+name should get prefixed with `wrhs_` so that it's less likely to conflict with
+an existing table name.
+
+The following sections contain the column/attribute mappings for each table.
+
+Legend:
+
+* **(pk)** - partition-key
+* **(ck)** - clustering key (Cassandra only)
+* **(sk)** - sort key (DynamoDB only)
+* **--** - Not in that table
+
+### Build (`build`)
+
+> **Open question:** With how DynamoDB works, should we just make the
```

Comment:
It's not something you can change after the fact without making a new table and migrating the data. adding additional attributes you can do on the fly, but not changing what the partition & sort keys are.  if we do `name` as the partition key we'd have `build_id` as the sort key, so to get a single build you query with both, if you want all builds for a package you queryAll for just the partition key. and if you want something in between you can do things with the sort key query that's effectively stuff like `startsWith 'foo!bar'`

The question here really gets boils down to performance and whats going to be fast to read for our 80% case. And I'm not sure we know that right now. Especially with sharding and having data at scale (though I'm not sure we'll have enough data in the near term for DynamoDB to start visibly sharding)

Reply:
My gut says to move forward how I detailed below, which more mimics how we had things in Cassandra and then monitor things.  If we have to do another migration to change the schema, hopefully we catch that sooner rather than later, but it's not the end of the world. Just more major version bumps. 😢